File: python/annotations/migration_states/migration_states.py
# ============================================================================
#  migration_states -- Python (ctypes) port of
#  examples\Vb6\annotations\migration_states\migration_states.bas
#  A square annotation whose review state is set to Completed then Accepted.
# ============================================================================
import os
import sys
# the wrapper lives at <root>/wrappers/python (source checkout) or beside
# the example tree (shipped package) -- find it without a hard-coded path
_d = os.path.dirname(os.path.abspath(__file__))
for _ in range(6):
    for _c in (os.path.join(_d, 'wrappers', 'python'), _d):
        if os.path.isfile(os.path.join(_c, 'lumaspdf.py')):
            sys.path.insert(0, _c)
            break
    else:
        _d = os.path.dirname(_d)
        continue
    break
import lumaspdf as L

NO_COLOR = 0xFFFFFFF1  # transparent


@L.TErrorProc
def err_proc(data, errcode, errmsg, errtype):
    return 0


def main():
    pdf = L.pdfNewPDF()
    L.pdfSetOnErrorProc(pdf, 0, err_proc)
    L.pdfCreateNewPDFA(pdf, b"")

    L.pdfSetPageCoords(pdf, L.pcTopDown)

    L.pdfAppend(pdf)
    # To see the migration state, right click on the annotation and then on Review History.
    annot = L.pdfSquareAnnotA(pdf, 50.0, 50.0, 200.0, 100.0, 3.0, NO_COLOR, 255, L.csDeviceRGB, b"Jim", b"Test", b"Just test...")
    reply = L.pdfSetAnnotMigrationStateA(pdf, annot, L.asCompleted, b"Harry")
    L.pdfSetAnnotStringA(pdf, reply, L.asContent, b"The state was set to Completed!")

    reply = L.pdfSetAnnotMigrationStateA(pdf, reply, L.asAccepted, b"Jim")
    L.pdfSetAnnotStringA(pdf, reply, L.asContent, b"The state was set to Accepted!")
    L.pdfEndPage(pdf)

    if L.pdfHaveOpenDoc(pdf) != 0:
        out_file = os.path.join(os.path.dirname(os.path.abspath(__file__)), "out.pdf")
        if L.pdfOpenOutputFileA(pdf, out_file.encode("latin-1")) == 0:
            L.pdfDeletePDF(pdf)
            return
        if L.pdfCloseFile(pdf) != 0:
            print('PDF file "' + out_file + '" successfully created!')

    L.pdfDeletePDF(pdf)


if __name__ == "__main__":
    main()

File: c/annotations/migration_states/migration_states.c
/* migration_states -- C (x64) port of examples\Vb6\annotations\migration_states.bas */
#include <stdio.h>
#include <string.h>
#include "lumaspdf.h"

#define NO_COLOR 0xFFFFFFF1u   /* transparent */

static SI32 PDF_CALL ErrProc(void* Data, SI32 ErrCode, const char* ErrMessage, SI32 ErrType)
{
    (void)Data; (void)ErrCode; (void)ErrType;
    if (ErrMessage) printf("%s\n", ErrMessage);
    return 0;
}

static void exedir(const char* a0, char* out, size_t n){char* s;strncpy(out,a0,n-1);out[n-1]=0;s=strrchr(out,'\\');if(!s)s=strrchr(out,'/');if(s)*s=0;else strcpy(out,".");}

int main(int argc, char** argv)
{
    SI32 annot, reply;
    PPDF pdf;
    char dir[1024], outFile[1200];

    exedir(argv[0], dir, sizeof(dir));
    (void)argc;

    pdf = pdfNewPDF();
    pdfSetOnErrorProc(pdf, 0, ErrProc);
    pdfCreateNewPDFA(pdf, "");

    pdfSetPageCoords(pdf, pcTopDown);

    pdfAppend(pdf);
    annot = pdfSquareAnnotA(pdf, 50.0, 50.0, 200.0, 100.0, 3.0, NO_COLOR, 255, csDeviceRGB, "Jim", "Test", "Just test...");
    reply = pdfSetAnnotMigrationStateA(pdf, annot, asCompleted, "Harry");
    pdfSetAnnotStringA(pdf, reply, asContent, "The state was set to Completed!");

    reply = pdfSetAnnotMigrationStateA(pdf, reply, asAccepted, "Jim");
    pdfSetAnnotStringA(pdf, reply, asContent, "The state was set to Accepted!");
    pdfEndPage(pdf);

    if (pdfHaveOpenDoc(pdf) != 0) {
        _snprintf(outFile, sizeof(outFile), "%s\\out.pdf", dir);
        if (pdfOpenOutputFileA(pdf, outFile) == 0) {
            pdfDeletePDF(pdf);
            return 0;
        }
        if (pdfCloseFile(pdf) != 0)
            printf("PDF file \"%s\" successfully created!\n", outFile);
    }

    pdfDeletePDF(pdf);
    return 0;
}
